### ISSUE TITLE
Question about NMI calculation and LL-FPN implementation

### ISSUE THREAD

**Issue by katz51** [casiatao/UNIP#7]
Hi, thank you for sharing your great work. I found your paper very impressive and am currently trying to implement it.

I have a couple of questions:

1. Where can I find the code for the NMI calculation? Which file contains this part?

    def calculate_normalized_mutual_information(attention_weights):
    query_prob_sum = torch.sum(attention_weights, dim=-1, keepdims=True)  # (B, H, L, 1)
    query_prob = query_prob_sum / torch.sum(query_prob_sum, dim=-2, keepdims=True)  # (B, H, L, 1) 

    key_prob_sum = torch.sum(attention_weights, dim=-2, keepdims=True)  # (B, H, 1, L)
    key_prob = key_prob_sum / torch.sum(key_prob_sum, dim=-1, keepdims=True)  # (B, H, 1, L)
    
    H_query = - torch.sum(query_prob * torch.log2(query_prob), dim=(-1, -2))  # (B, H)
    H_key = - torch.sum(key_prob * torch.log2(key_prob), dim=(-1, -2)) # (B, H)

    qk_joint_prob = attention_weights * query_prob  # (B, H, L, L)
    I = torch.sum(qk_joint_prob * torch.log2(qk_joint_prob / (query_prob * key_prob)), axis=(-1, -2))  # (B, H)
    normalized_I = I / torch.sqrt(H_query * H_key)  # (B, H)
    normalized_I = torch.mean(normalized_I, dim=-1) # (B)
    return normalized_I

2. In the paper, it seems that only Block12 is used as input to UperNet via LL-FPN. However, in the code, multiple feature maps (e.g., four blocks specified by out_indices) are used as inputs. Could you clarify where the LL-FPN is implemented in the code?

Thanks in advance for your help!
